src/DS/metrics_extraction.py

```diff
@@ -0,0 +1,407 @@
+
+Notes about the input and output files of this script:
+- Input file : This is the output of the Supabase database after running the
+script `guesslang_to_db.py`. This dataframe contains the repository `file_
+path` and 'guessed_language' amongst other columns like `hosting_provider` and
+`repo_name`.
+- output file: is a dataframe which contains information about testing
+techniques and code complexity. Some columns are `num_assertions`,
+`cyclomatic_complexity`, and `lines_of_code`.
+
+### Usage:
+To run the script, specify the input and output CSV file paths using the
+command-line arguments `--input` and `--output`. By default, the script
+assumes input and output files are located in the 'data' directory of the
+repository's root.
+
+Example:
+    python script_name.py --input path/to/input.csv --output path/to/output.csv
+
+"""
+
+import argparse
+import ast
+import pandas as pd
+from mccabe import PathGraphingAstVisitor
+from loguru import logger
+from pathlib import Path
+
+from utils.git_utils import get_working_directory_or_git_root
+
+BATCH_SIZE = 100  # Number of files to process before saving to disk
+
+# Configure logger
+logger.add("metrics_extraction.log", rotation="500 MB", level="INFO")
+
+
+def parse_args():
+    parser = argparse.ArgumentParser(
+        description="Analyse Python test files and general Python code files "
+        "and extract metrics."
+    )
+
+    parser.add_argument(
+        "--input",
+        type=str,
+        default=Path(
+            get_working_directory_or_git_root() / "data" / "guessed_languages_rows.csv"
+        ),
+        help="Path to the input CSV file exported from Supabase with "
+        "file_path, guessed_language, and more.",
+    )
+
+    parser.add_argument(
+        "--output",
+        type=str,
+        default=Path(
+            get_working_directory_or_git_root()
+            / "data"
+            / "test_metrics_df_guesslang.csv"
+        ),
+        help="Path to the output CSV file with data on testing techniques and "
+        "code complexity.",
+    )
+
+    return parser.parse_args()
+
+
+def read_and_parse_file(file_path):
+    """
+    Reads and parses a Python file into an AST.
+
+    Args:
+        file_path (str): The path to the Python file.
+
+    Returns:
+        tuple: A tuple containing the AST tree and the file content, or
+        None if there's an error.
+    """
+    logger.debug(f"Attempting to read and parse: {file_path}")
+    if not file_path.endswith(".py"):
+        logger.info(f"Skipping non-python file {file_path}")
+        return None
+
+    try:
+        with open(file_path, "r", encoding="utf-8") as file:
+            content = file.read()
+            tree = ast.parse(content, filename=file_path)
+            logger.debug(f"Successfully parsed file: {file_path}")
+            return tree, content
+    except SyntaxError as e:
+        logger.error(f"Syntax error in file {file_path}: {e}")
```

Comment:
It might be worth a chat about whether these are better flagged as errors or warnings in the log... 

One option is to only treat errors in this program as errors from a logging perspective and to treat errors/problems in the contents of files as warnings for someone to consider investigating. _c.f._ the distinction between `4XX` and `5XX` classes of errors in HTTP response codes - which ones can be corrected by the caller and which need to be corrected by the service provider.